### PR TITLE
Feat/log on invalid cowboy version

### PR DESCRIPTION
**Pull request by Gazler**
This relies on the Cowboy 2 PR, so that should be merged first. I discovered this could happen quite easily when upgrading an app using the gr-cowboy2 branch.

If a user updates their cowboy dependency without updating the relevant
hander in their config then the warning will be displayed. This should
help minimize issues when users simply update the Cowboy version without
also updating the config.

### CODE REVIEW

**Comment by josevalim** [comment]
I think we should go ahead and autodetect the handler. Basically:

1. autodetect the handler to see if we are using cowboy1 or cowboy2
2. if the user gave a handler, warn if user_handler is cowboy1 and autodetect is cowboy2
3. then return `user_handler || autodetected_handler`

Thoughts?

**Comment by Gazler** [comment]
@josevalim sounds good to me. Should I keep the log message the same as in this PR?

File: lib/phoenix/endpoint/cowboy_handler.ex [hunk 1]
```diff
@@ -114,4 +116,23 @@ defmodule Phoenix.Endpoint.CowboyHandler do
...
+        Logger.warn("""
+        This handler is for Cowboy 1.x and you are using Cowboy #{version}
+
+        If you wish to use Cowboy 1.x, please update mix.exs to point to the
+        correct Cowboy version. If you want to use Cowboy 2, then please set:
+
+            handler: Phoenix.Endpoint.Cowboy2Handler,
+
+        In your config.exs file.
+
```
Comment (josevalim) [inline lib/phoenix/endpoint/cowboy_handler.ex]: I think you don't need this trailing newline.

File: lib/phoenix/endpoint/cowboy_handler.ex [hunk 2]
```diff
@@ -114,4 +116,23 @@ defmodule Phoenix.Endpoint.CowboyHandler do
     addr_str = :inet.ntoa(addr)
     "Running #{inspect endpoint} with Cowboy using #{scheme}://#{addr_str}:#{port}"
   end
+
+  defp log_on_invalid_cowboy_version(endpoint) do
+    case Application.spec(:cowboy, :vsn) do
+      [?1 | _] -> nil
+      version ->
+        Logger.warn("""
+        This handler is for Cowboy 1.x and you are using Cowboy #{version}
```
Comment (josevalim) [inline lib/phoenix/endpoint/cowboy_handler.ex]: "This handler" is not clear. Probably something like this:

```
You have specified Phoenix.Handlers.Cowboy for cowboy v1.x in the
:handler configuration of your Phoenix endpoint #{inspect endpoint}
but your mix.exs has fetched cowboy v2.x.
```

The rest of the message is perfect.

File: lib/phoenix/endpoint/cowboy_handler.ex [hunk 3]
```diff
@@ -114,4 +116,23 @@ defmodule Phoenix.Endpoint.CowboyHandler do
...
+        correct Cowboy version. If you want to use Cowboy 2, then please set:
+
+            handler: Phoenix.Endpoint.Cowboy2Handler,
+
+        In your config.exs file.
+
+        """)
+    end
+  end
+
```
Comment (chrismccord) [inline lib/phoenix/endpoint/cowboy_handler.ex]: ✂️

**Comment by Gazler** [comment]
@josevalim I've updated this to autodetect. It is very unlikely that the CowboyHandler can be used when the Cowboy version is 2.x - so maybe we should raise instead of warning? The only situation I can think of is that someone has their own fork which has a different non-1.x version for Cowboy.

**Comment by Gazler** [comment]
@josevalim @chrismccord Are you both happy with this approach?

**Comment by josevalim** [comment]
@Gazler I think we should raise in all cases, since it will certainly fail.

**Pull request merged**